### PR TITLE
Add a OneOneMatching distribution

### PR DESCRIPTION
**Pull request by fritzo**
This follows #2697 by introducing a `OneOneMatching` distribution that is nearly identical to, but slightly simpler than, the earlier `OneTwoMatching`. (Indeed this PR took only around 15 minutes to create. @martinjankowiak I hope it will be similarly easy to review 😬 )

My motivation is (1) completeness, (2) to have a simpler set of tests where I can experiment with better BP algorithms, and (3) to verify the relationship between 1-1-matchings and 1-2-matchings.

## Tested
- [x] added tests as in the original PR
- [x] added gradcheck tests for `OneOneMatching`
- [x] added gradcheck tests for `OneTwoMatching`

### CODE REVIEW

File: tests/distributions/test_one_one_matching.py [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+    with xfail_if_not_implemented():
+        values = d.sample(sample_shape)
+    assert values.shape == sample_shape + (num_nodes,)
+    assert d.support.check(values).all()
+
+
+@pytest.mark.parametrize("dtype", [torch.float, torch.double], ids=str)
+@pytest.mark.parametrize("num_nodes", [1, 2, 3, 4, 5, 6, 7, 8])
+@pytest.mark.parametrize("bp_iters", [None, BP_ITERS], ids=["exact", "bp"])
+def test_log_prob_full(num_nodes, dtype, bp_iters):
```
Comment (martinjankowiak) [inline tests/distributions/test_one_one_matching.py:47]: do you think it would be useful to pair one of these with a finite difference gradient test?

Comment (fritzo) [reply tests/distributions/test_one_one_matching.py:47]: good idea. I've added gradient check tests for both `OneOneMatching` and `OneTwoMatching`. Note the tolerances are pretty coarse (atol=rtol=0.01) and I needed to run for a few more iterations.

File: pyro/distributions/one_one_matching.py [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+            warnings.warn("Matching is not binary")
+            return torch.tensor(False)
+        return torch.tensor(True)
+
+
+class OneOneMatching(TorchDistribution):
+    r"""
+    Random perfect matching from ``N`` sources to ``N`` destinations where each
+    source matches exactly **one** destination and each destination matches
+    exactly **one** sources.
```
Comment (martinjankowiak) [inline pyro/distributions/one_one_matching.py:45]: nit: typo sources

File: tests/distributions/test_one_one_matching.py [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+    i = torch.arange(num_nodes)
+    logits = torch.randn(num_nodes, num_nodes) * 10
+    logits[i[:, None] < i] = -100
+    logits.requires_grad_()
+
+    def fn(logits):
+        d = dist.OneOneMatching(logits, bp_iters=bp_iters)
+        return d.log_partition_function
+
+    torch.autograd.gradcheck(fn, logits, atol=0.01, rtol=0.01)
```
Comment (martinjankowiak) [inline tests/distributions/test_one_one_matching.py:95]: great these are reassuring!

**Review by martinjankowiak — DISMISSED** [review]
how expensive are these tests? quite a good number with all the `mark.parametrize`

**Comment by fritzo** [comment]
> how expensive are these tests?

It takes 15 seconds to run all the matching tests on my laptop (in a single process).

<details>

```
% time pytest tests/distributions/test_one_*_matching.py
=============================================== test session starts ===============================================
platform darwin -- Python 3.7.0, pytest-6.1.2, py-1.9.0, pluggy-0.13.1
benchmark: 3.2.3 (defaults: timer=time.perf_counter disable_gc=False min_rounds=5 min_time=0.000005 max_time=1.0 calibration_precision=10 warmup=False warmup_iterations=100000)
rootdir: /Users/fobermey/github/pyro-ppl/pyro, configfile: setup.cfg
plugins: forked-1.2.0, nbval-0.9.6, xdist-2.1.0, benchmark-3.2.3
collected 400 items

tests/distributions/test_one_one_matching.py ................................................xxxxxxxxxxxxxx [ 15%]
xxxxxxxxxxxxxxxxxxxxxx..................................................................................... [ 42%]
.....xxxxxxxxxx                                                                                             [ 46%]
tests/distributions/test_one_two_matching.py ........................................xxxxxxxxxxxxxxxxxxxxxx [ 61%]
xxxxxxxx................................................................................................... [ 88%]
.............................xxxxxx......xxxxxx                                                             [100%]

======================================== 312 passed, 88 xfailed in 14.40s =========================================
pytest tests/distributions/test_one_*_matching.py  18.20s user 2.24s system 128% cpu 15.862 total
```

</details>

**Review by martinjankowiak — APPROVED** [review]

**Comment by fritzo** [comment]
Thanks for reviewing @martinjankowiak !
